ksqldb-engine/src/main/java/io/confluent/ksql/analyzer/AggregateAnalyzer.java

```diff
@@ -128,48 +137,61 @@ private void processSelect(final Expression expression) {
     }
 
     private void processGroupBy(final Expression expression) {
-      final AggregateVisitor visitor = new AggregateVisitor(this, (aggFuncName, node) -> {
-        if (aggFuncName.isPresent()) {
-          throw new KsqlException("GROUP BY does not support aggregate functions: "
-              + aggFuncName.get().text() + " is an aggregate function.");
-        }
-        throwOnWindowBoundColumnIfWindowedAggregate(node);
-      });
+      final AggregateVisitor visitor = new AggregateVisitor(
+          this,
+          groupBy,
+          foundExpressionInGroupBy,
+          (aggFuncName, node) -> {
+            if (aggFuncName.isPresent()) {
+              throw new KsqlException("GROUP BY does not support aggregate functions: "
+                  + aggFuncName.get().text() + " is an aggregate function.");
+            }
+            throwOnWindowBoundColumnIfWindowedAggregate(node);
+          });
 
       visitor.process(expression, null);
     }
 
     private void processWhere(final Expression expression) {
-      final AggregateVisitor visitor = new AggregateVisitor(this, (aggFuncName, node) ->
-          throwOnWindowBoundColumnIfWindowedAggregate(node));
+      final AggregateVisitor visitor = new AggregateVisitor(
+          this,
+          groupBy,
+          foundExpressionInGroupBy,
+          (aggFuncName, node) ->
+            throwOnWindowBoundColumnIfWindowedAggregate(node));
 
       visitor.process(expression, null);
     }
 
     private void processHaving(final Expression expression) {
-      final AggregateVisitor visitor = new AggregateVisitor(this, (aggFuncName, node) -> {
-        throwOnWindowBoundColumnIfWindowedAggregate(node);
-
-        if (!aggFuncName.isPresent()) {
-          captureNoneAggregateHavingNotPartOfGroupBy(node);
-        }
-      });
+      final AggregateVisitor visitor = new AggregateVisitor(
+          this,
+          groupBy,
+          foundExpressionInGroupBy,
+          (aggFuncName, node) -> {
+            throwOnWindowBoundColumnIfWindowedAggregate(node);
+
+            if (!aggFuncName.isPresent()) {
+              captureNoneAggregateHavingNotPartOfGroupBy(node);
+            }
+          });
 
       visitor.process(expression, null);
 
       aggregateAnalysis.setHavingExpression(expression);
     }
 
-    private void throwOnWindowBoundColumnIfWindowedAggregate(
-        final ColumnReferenceExp node
-    ) {
+    private void throwOnWindowBoundColumnIfWindowedAggregate(final Expression node) {
       // Window bounds are supported for operations on windowed sources
       if (!analysis.getWindowExpression().isPresent()) {
         return;
       }
 
       // For non-windowed sources, with a windowed GROUP BY, they are only supported in selects:
-      if (SystemColumns.isWindowBound(node.getColumnName())) {
+      if (!(node instanceof ColumnReferenceExp)) {
+        return;
+      }
```

Comment:
The comment on line 190 refers to code now on line 194.  Can you move this `if` statement above the comment please?